third/jwgc/clients/jstat/jstat.1
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "JSTAT 1"
.TH JSTAT 1 "2003-10-06" "beta4" "Jabber Windowgram Client"
.SH "NAME"
jstat \- display Jwgc status
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
jstat [ \-h ] [ \-d <flags> ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIJstat\fR is used to display the current status of \fIjwgc\fR's connection to
the Jabber server.  It lists out the services available from the Jabber
server you are connected to, some information about your connection
to the Jabber server, and some various statistics.
.PP
\&\fIJstat\fR understands the following options:
.PP
\&\-h             Displays usage information.
.PP
\&\-d <flags>     Debugging output.  Run without <flags> for usage.     
.SH "AUTHOR"
.IX Header "AUTHOR"
Daniel Henninger <jadestorm@users.sourceforge.net>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2002 Daniel Henninger. All right reserved.
This program is free software; you can redistribute it
and/or modify it under the terms explained in the \fB\s-1LICENSE\s0\fR
file that comes with the \fIjwgc\fR distribution.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIjwgc\fR\|(1), \fIjwrite\fR\|(1), \fIjctl\fR\|(1), \fIjlocate\fR\|(1)


third/jwgc/clients/jstat/jstat.c
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <libjwgc.h>

char *whoami;

void
usage()
{
	fprintf(stderr, "Usage: %s %s[-h]\n%s\
  -h       Display help\n\
",
	whoami,
#ifdef NODEBUG
	"",
	""
#else
	"[-d <flags>] ",
	"  -d       Enable/Disable debugging (leave <flags> blank for usage)\n"
#endif /* NODEBUG */
	);
	exit(1);
}

void
jstat_on_event_handler(jwgconn conn, jwgpacket packet)
{
	xode x, y;
	char *cdata;

	x = packet->x;
	cdata = xode_get_tagdata(x, "user");
	if (cdata) {
		printf("User:               %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "server");
	if (cdata) {
		printf("Server:             %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "resource");
	if (cdata) {
		printf("Resource:           %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "version");
	if (cdata) {
		printf("Jwgc Version:       %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "machinetype");
	if (cdata) {
		printf("Machine Type:       %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "localtime");
	if (cdata) {
		printf("Localhost Time:     %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "connecttime");
	if (cdata) {
		printf("Connection Time:    %.58s\n", cdata);
	}

	cdata = xode_get_tagdata(x, "connectstate");
	if (cdata) {
		printf("Connection State:   %.58s\n", cdata);
	}

	printf("\nServices Available:\n");
	y = xode_get_tag(x, "agents");
	y = xode_get_firstchild(y);
	while (y) {
		char *name, *service, *jid;
		xode z;

		name = xode_get_attrib(y, "name");
		service = xode_get_attrib(y, "service");
		jid = xode_get_attrib(y, "jid");
		printf("    %-30.30s %41.41s\n", name, jid);
		printf("          Provides: %s", service);
		z = xode_get_firstchild(y);
		while (z) {
			printf(" %s", xode_get_name(z));
			z = xode_get_nextsibling(z);
		}
		printf("\n");
		y = xode_get_nextsibling(y);
	}

	cdata = xode_get_tagdata(x, "bugreport");
	if (cdata) {
		printf("\nPlease report bugs to %s.\n", cdata);
	}
}


int
main(argc, argv)
	int argc;
	char *argv[];
{
	int arg;
	xode x;
	jwgconn jwg;

	whoami = argv[0];
	dinit();

	arg = 1;

	for (; arg < argc; arg++) {
		switch (argv[arg][1]) {
#ifndef NODEBUG
			case 'd':
				arg++;
				if (arg >= argc) {
					dprinttypes();
				}
				dparseflags(argv[arg]);
				break;
#endif /* NODEBUG */
			case 'h':
				usage();
				break;
			default:
				fprintf(stderr, "Unknown option: %s\n",
					argv[arg]);
				usage();
		}
	}

	jwg = jwg_new();
	if (!jwg) {
		fprintf(stderr, "jstat: failed to initialize jwgc connection\n");
		exit(1);
	}
	jwg_event_handler(jwg, jstat_on_event_handler);
	jwg_start(jwg);
	if (jwg_getfd(jwg) < 0) {
		fprintf(stderr, "jstat: failed to create jwgc connection\n");
		exit(1);
	}
	x = xode_new("status");
	jwg_send(jwg, x);
	xode_free(x);

	jwg_poll(jwg, -1);
	jwg_stop(jwg);

	exit(0);
}
